pkg/db/queries.sql

```diff
@@ -13,23 +13,36 @@ WHERE
 	singleton_id = 1;
 
 -- name: InsertGatewayEnvelope :execrows
-SELECT
-	insert_gateway_envelope(@originator_node_id, @originator_sequence_id, @topic, @originator_envelope);
+INSERT INTO gateway_envelopes(originator_node_id, originator_sequence_id, topic, originator_envelope)
+	VALUES (@originator_node_id, @originator_sequence_id, @topic, @originator_envelope)
+ON CONFLICT
+	DO NOTHING;
 
 -- name: SelectGatewayEnvelopes :many
+WITH cursors AS (
+	SELECT
+		UNNEST(@cursor_node_ids::INT[]) AS cursor_node_id,
+		UNNEST(@cursor_sequence_ids::BIGINT[]) AS cursor_sequence_id
+)
 SELECT
-	*
+	gateway_envelopes.*
 FROM
 	gateway_envelopes
+	-- Assumption: There is only one cursor per node ID. Caller must verify this
+	LEFT JOIN cursors ON gateway_envelopes.originator_node_id = cursors.cursor_node_id
```

Comment:
The most important changes are in this query. We're doing this interesting JOIN operation against the cursor specified by the caller.

I originally wanted to use a different approach of specifying one WHERE clause per cursor entry, but as far as I can tell [sqlc does not support this](https://github.com/sqlc-dev/sqlc/issues/3578).

The JOIN method is more readable IMO, although I'm unsure of the performance implication - it depends on how much optimization postgres does internally. When we are further along in the project, we can do some benchmarking with `explain analyze` etc.

Reply:
If it turns out to be a big deal, there is always the option of just constructing the parameterized SQL statement outside of sqlc and running it against the DB driver.